source/common/stats/stats_impl.cc

```diff
@@ -256,18 +272,23 @@ TagProducerImpl::addDefaultExtractors(const envoy::config::metrics::v2::StatsCon
 }
 
 void HeapRawStatDataAllocator::free(RawStatData& data) {
-  // This allocator does not ever have concurrent access to the raw data.
-  ASSERT(data.ref_count_ == 1);
+  ASSERT(data.ref_count_ > 0);
+  if (--data.ref_count_ > 0) {
+    return;
+  }
+
+  size_t key_removed = stats_set_.erase(std::string(data.key()));
```

Comment:
We expect all stats to call freed before the allocator object disappears, so I would suggest adding a destructor with an `ASSERT` to check that the map is empty.